### PR TITLE
Replace dlfcn-win32 with simple-dlfcn-win32

### LINKED ISSUE #223: node-ffi claims MIT license, but deps/dlfcn-win32 uses LGPL
The licensing situation on Windows is mixed. node-ffi's own license is MIT, but on Windows it depends on dlfcn-win32, which is LGPL.

I would suggest either

(1) Ditch dnfcn-win32 and use native Win32 DLL API's directly. IMHO, trying to emulate POSIX semantics on Windows is a mistake. Windows has no concept of a global symbol table, and multiple modules may be loaded with conflicting exports. In particular, multiple C runtime libraries are common.

or

(2) Update node-ffi's LICENSE file to include note about dlfcn-win32 using LGPL. Some organizations prohibit the use of [L]GPL code, or require legal review to do so. It would be helpful to warn them.

### PR DESCRIPTION
**Pull request by mcnameej**
I'm submitting two PR's to fix #223. Pick one. This PR replaces the dlfcn-win32 code.

### CODE REVIEW

**Comment by unbornchikken** [comment]
I'd go with this definitely, but in this case, we dont need this reside in deps. Could you integrate it with src?

Thanks a lot!

**Comment by mcnameej** [comment]
Three test cases fail with my version of dlfcn...

```
1) Library should accept `null` as a first argument:
   Error: Dynamic Symbol Retrieval Error: Win32 error 127
    at DynamicLibrary.get (C:\Proj\iojs\node-ffi\lib\dynamic_library.js:112:11)
    at C:\Proj\iojs\node-ffi\lib\library.js:50:19
    at Array.forEach (native)
    at new Library (C:\Proj\iojs\node-ffi\lib\library.js:47:28)
    at Context.<anonymous> (C:\Proj\iojs\node-ffi\test\library.js:24:21)

2) Library should work with "strcpy" and a 128 length string:
   Error: Dynamic Symbol Retrieval Error: Win32 error 127
    at DynamicLibrary.get (C:\Proj\iojs\node-ffi\lib\dynamic_library.js:112:11)
    at C:\Proj\iojs\node-ffi\lib\library.js:50:19
    at Array.forEach (native)
    at new Library (C:\Proj\iojs\node-ffi\lib\library.js:47:28)
    at Context.<anonymous> (C:\Proj\iojs\node-ffi\test\library.js:64:18)

3) Library should work with "strcpy" and a 2k length string:
   Error: Dynamic Symbol Retrieval Error: Win32 error 127
    at DynamicLibrary.get (C:\Proj\iojs\node-ffi\lib\dynamic_library.js:112:11)
    at C:\Proj\iojs\node-ffi\lib\library.js:50:19
    at Array.forEach (native)
    at new Library (C:\Proj\iojs\node-ffi\lib\library.js:47:28)
    at Context.<anonymous> (C:\Proj\iojs\node-ffi\test\library.js:74:18)
```

These test cases depend on POSIX semantics in dlsym: searching all loaded modules to find a symbol by name.  No Windows programmer would ever think to do that.

When you search for "strcpy", which strcpy() do you expect to find?  ntdll.dll is loaded into all processes, and it exports strcpy.  msvcrt.dll is loaded into most processes, but it belongs to the OS -- node.exe doesn't use it.  In fact, node.exe static links it's C runtime, and those symbols are unavailable (even the original dlfcn-win32, which tries to emulate POSIX, can't see them).  If node.exe did use a C runtime DLL, it would be something like msvcr120.dll.  You have several DLL's loaded in your process that export "strcpy".  Which one will dlsym return?  Who knows.

It probably doesn't matter which runtime you get strcpy() from.  It matters a lot for things like malloc(), free(), fopen(), and fclose().  Memory allocated by malloc() must be passed to the free() **in the same C runtime** as the malloc() that allocated it.  The FILE* returned by fopen() can only be passed to stdio functions in the same C runtime.  This is why Windows DLL's generally don't use stdio FILE's in their public interface, and if they have functions that return an allocated buffer, they'll provide a function to free it (of they'll allocate the memory using a Win32 API like GlobalAlloc() or HeapAlloc(), which any caller can free).

**Comment by unbornchikken** [comment]
I understand your point, and agree. I think those test are not there to test strcmp, those are there to test if binding is works or not by the given situation. If you could replace them with an invocations of an strcmp wrapper method in ffi_test.node, then I'd accept it that way. That node file is just a dynamic lib, so it gotta work that way. @TooTallNate waddaya say?

**Comment by mcnameej** [comment]
P.S. My last commit included a Unicode test case similar to PR #225.

**Comment by unbornchikken** [comment]
So to sum it up:

- finally we can drop another dependency, only ffi remains
- we have a simplifed dlopen and co. on Windows that also fixes https://github.com/node-ffi/node-ffi/issues/183
- there are some updated test cases of testing the same node-ffi behavior, but wouldn't encourage anyone to use node-ffi in a way that really never works right on Windows

@TooTallNate that's what I call pull request. Could be merged ASAP. Agree?

**Comment by unbornchikken** [comment]
I celebrated too early. Unfortunattely we have some failing test cases by merging this on Linux. Please take a look at the Travis log. Do you have experience to fix them? If not, I can help you with this.

**Comment by mcnameej** [comment]
I'll fire up a Linux VM and investigate the failures reported by travis.

Not sure what to make of the failures reported by AppVeyor.  Is that a problem with the patch, or a problem with the environment?

**Comment by unbornchikken** [comment]
You can ignore AppVeyor's log, thats because of iojs.exe instead of node.exe issue. Travis should work with 0.10 and 0.12. With io.js it will fail because 3.x support is blocked because of ref module is not yet implemented its NAN 2.0 support.

File: src/win32-dlfcn.cc [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        sprintf(errorMessage, "Win32 error %lu", lastError);
+        lastError = 0;
+        return errorMessage;
+    } else {
+        return NULL;
+    }
+
+#else
+
+    /* "Ken Thompson has an automobile which he helped design..." */
```
Comment (TooTallNate) [inline src/win32-dlfcn.cc]: Why is this part necessary out of curiosity?

**Comment by TooTallNate** [comment]
I'm :+1: for this, but ya we'll have to make sure all the tests are in order. It could be technically considered a breaking change for Windows users who might have been depending on the previous `null` behavior, but I'm contemplating whether we can consider it a bug fix instead since the behavior was _incorrect_ from the get-go.

File: src/win32-dlfcn.cc (hunk 1 above)
Comment (mcnameej) [inline src/win32-dlfcn.cc]: Depends how you define "necessary".

I think dlerror() is an extremely bad design, but it is a POSIX standard.  The question becomes how hard to you want to work to implement it on Windows.

I'd be happy with my original implementation (always returning "?"), because IMHO, nobody should be relying on dlerror() to begin with.

The current implementation provides standard-compliant dlerror() functionality  by returning a string with the Win32 error code (which can be looked up in WinError.h if ever needed).

What I didn't want to do is return the actual Win32 error text.  Win32 error messages can be verbose.  I'd either have to allocate a buffer that nobody will free (see also: memory leak), or create a large static buffer.  For what?  A message nobody should be relying on.

**Comment by unbornchikken** [comment]
Cool, it's finally ready. I try to make a NAN 2.0 support PR to ref module on this week, and travis is gonna be happy again.

File: src/win32-dlfcn.cc (hunk 1 above)
Comment (TooTallNate) [inline src/win32-dlfcn.cc]: Well the basic implementation above seems adequate to me, and with the `#if` above this is dead code. Maybe remove to avoid future confusion?

**Comment by unbornchikken** [comment]
I'll merge this once NAN 2.0 PR gets ready (after this gets landed: https://github.com/TooTallNate/ref/pull/33), and we can release those together with all tests passed on Travis.

**Comment by TooTallNate** [comment]
I have this merged in the `v2` branch. v0.10 seems to be failing one test. See https://ci.appveyor.com/project/TooTallNate/node-ffi/build/80. @mcnameej Any ideas there?

```
[00:01:36]   1) Library should accept `null` as a first argument:
[00:01:36]      Error: Dynamic Symbol Retrieval Error: Win32 error 127
[00:01:36]       at DynamicLibrary.get (C:\projects\node-ffi\lib\dynamic_library.js:112:11)
[00:01:36]       at C:\projects\node-ffi\lib\library.js:50:19
[00:01:36]       at Array.forEach (native)
[00:01:36]       at new Library (C:\projects\node-ffi\lib\library.js:47:28)
[00:01:36]       at Context.<anonymous> (C:\projects\node-ffi\test\library.js:33:22)
[00:01:36]       at callFn (C:\projects\node-ffi\node_modules\mocha\lib\runnable.js:289:21)
[00:01:36]       at Test.Runnable.run (C:\projects\node-ffi\node_modules\mocha\lib\runnable.js:282:7)
[00:01:36]       at Runner.runTest (C:\projects\node-ffi\node_modules\mocha\lib\runner.js:419:10)
[00:01:36]       at C:\projects\node-ffi\node_modules\mocha\lib\runner.js:526:12
[00:01:36]       at next (C:\projects\node-ffi\node_modules\mocha\lib\runner.js:340:14)
[00:01:36]       at C:\projects\node-ffi\node_modules\mocha\lib\runner.js:350:7
[00:01:36]       at next (C:\projects\node-ffi\node_modules\mocha\lib\runner.js:283:14)
[00:01:36]       at Object._onImmediate (C:\projects\node-ffi\nod
[00:01:36] e_modules\mocha\lib\runner.js:318:5)
```

**Comment by mcnameej** [comment]
@TooTallNate: The problem was that older versions of node.exe don't export a ``node_module_register`` function.  I changed the test to use ``uv_fs_open`` instead.  PR #232 contains the fix for v2 branch.

**Comment by TooTallNate** [comment]
Merged. Thanks @mcnameej!
